### PR TITLE
Chore: Reduce number of strict typescript errors

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
Fixes a number of strict typescript errors as part of ongoing efforts.

### CODE REVIEW

**Review by jackw — APPROVED** [review]
LGTM! 👍

File: public/app/features/explore/Logs.tsx [hunk 1]
```diff
@@ -80,8 +80,8 @@ interface State {
 }
 
 export class UnthemedLogs extends PureComponent<Props, State> {
-  flipOrderTimer: NodeJS.Timeout;
-  cancelFlippingTimer: NodeJS.Timeout;
+  flipOrderTimer?: NodeJS.Timeout;
+  cancelFlippingTimer?: NodeJS.Timeout;
```
Comment (dprokop) [inline public/app/features/explore/Logs.tsx:84]: you can get rid of NodeJS annotations and replace them with number if wou change setTimeout/clearTimeout to window.setTimeout/window.clearTimeout

File: public/app/plugins/panel/graph/Legend/LegendSeriesItem.tsx [hunk 2]
```diff
@@ -4,7 +4,8 @@ import { TimeSeries } from 'app/core/core';
 import { SeriesColorPicker, SeriesIcon } from '@grafana/ui';
 import { selectors } from '@grafana/e2e-selectors';
 
-export const LEGEND_STATS = ['min', 'max', 'avg', 'current', 'total'];
+export const LEGEND_STATS = ['min', 'max', 'avg', 'current', 'total'] as const;
```
Comment (dprokop) [inline public/app/plugins/panel/graph/Legend/LegendSeriesItem.tsx:7]: Maybe we could use enum values: https://github.com/grafana/grafana/blob/408bc06bab5efa2d5432a70c3d8079c7e2ef9ae5/packages/grafana-data/src/transformations/fieldReducer.ts#L7 instead?

**Review by dprokop — APPROVED** [review]
LGTM

File: public/app/core/navigation/parseKeyValue.ts [hunk 3]
```diff
@@ -123,6 +126,7 @@ export function tryDecodeURIComponent(value: string): string | void {
     return decodeURIComponent(value);
   } catch (e) {
     // Ignore any invalid uri component.
+    return;
```
Comment (hugohaggmark) [inline public/app/core/navigation/parseKeyValue.ts:129]: Maybe we should return an empty string here and then we could get rid of `isDefined` function all together? Also, I think we should remove the `isUndefined` function on line 163 which doesn't seem to be used. 

Something to consider outside of the scope of this PR: why are most of these functions exported? @dprokop any thoughts on this?

File: public/app/features/panel/metrics_panel_ctrl.ts [hunk 4]
```diff
@@ -19,12 +19,12 @@ import { PanelModel } from 'app/features/dashboard/state';
...
 
 class MetricsPanelCtrl extends PanelCtrl {
-  datasource: DataSourceApi;
+  datasource?: DataSourceApi;
   contextSrv: ContextSrv;
   datasourceSrv: any;
   timeSrv: any;
   templateSrv: any;
-  range: TimeRange;
+  range?: TimeRange;
```
Comment (hugohaggmark) [inline public/app/features/panel/metrics_panel_ctrl.ts:27]: Hmm, I wonder if this is a case of https://github.com/grafana/grafana/pull/33012#discussion_r613806336

File: public/app/features/panel/metrics_panel_ctrl.ts [hunk 5]
```diff
@@ -19,12 +19,12 @@ import { PanelModel } from 'app/features/dashboard/state';
 import { PanelQueryRunner } from '../query/state/PanelQueryRunner';
 
 class MetricsPanelCtrl extends PanelCtrl {
-  datasource: DataSourceApi;
+  datasource?: DataSourceApi;
```
Comment (hugohaggmark) [inline public/app/features/panel/metrics_panel_ctrl.ts:22]: I remember we discussed something about babel, Angular and marking props as undefined. But I don't recall right now what it was, I think props will not exist or something like that.

File: public/app/plugins/panel/heatmap/rendering.ts [hunk 6]
```diff
@@ -161,7 +161,7 @@ export class HeatmapRenderer {
     const ticks = this.chartWidth / DEFAULT_X_TICK_SIZE_PX;
     const format = graphTimeFormat(ticks, this.timeRange.from.valueOf(), this.timeRange.to.valueOf());
     const timeZone = this.ctrl.dashboard.getTimezone();
-    const formatter = (date: Date) =>
+    const formatter = (date: d3.AxisDomain) =>
```
Comment (hugohaggmark) [inline public/app/plugins/panel/heatmap/rendering.ts:164]: Maybe we should define our own type for this?

**Review by hugohaggmark — COMMENTED** [review]
Outstanding work taking on this, left some comments for consideration.

**Pull request merged**